### DISCUSSION TITLE
What would be the approach for customizing json encoding, or internal object representation?

### DISCUSSION THREAD

**Discussion by danielSanchezQ** [pydantic/pydantic#2561] [Question]
Hi! I'm working on implementing pydantic support for our library [`rusty_results`](https://github.com/danielSanchezQ/rusty_results)

I wanted our custom `Option` type (pretty much like typing `Optional` but on steroids) to behave transparently. Let me explain. Right now it is implemented as a generic dataclass. So the representation when serializing is a json object like `{"some": "value"}` or `{}` for empty case. What would be better would be for it to behave as `Optional` works right now (single value, or default None that is handled in the validation(?)). 

I struggled a lot but I couldn't find a way to go around the current systems.

I must say it may be complicated since we do not want pydantic to be a required dependency (hence we cannot make our dataclasses inherit from BaseModel for example).

I would love to have some help. Thanks in advance!

**Comment by danielSanchezQ** [comment]
I would be happy to create a PR if any change is needed for this (if it is not supported ATM and it is something to be considered). I would just need a bit of mentorship to point me in the right direction.

**Reply by PrettyWood** [reply]
Hi @danielSanchezQ 
I may be able to help since I've worked a lot on pydantic, dataclasses and use Rust on a daily basis.
But I don't understand why you talk about a PR on pydantic and at the same time say it won't be a required dependency.
Do you want to customize the serialization of `Option` and make it work automatically when used inside pydantic models?
Anyway I'd be happy to help if I can

**Reply by danielSanchezQ** [reply]
Hi @PrettyWood !

First, thanks for the reply!

> Do you want to customize the serialization of Option and make it work automatically when used inside pydantic models?

Yes, that is exactly what I want. It would be awesome if serizliation/deserialization would work as typing `Optional` works right now. I tried, but my experience with pydantic is quite basic atm.

I talked about making a PR, in case the library doesn't support such behaviour and it is interesting for pydantic ofc. I'm open to help too 😄 

So, any insight is more than welcome. Right now we support pydantic for our types with some basic outcome: https://github.com/danielSanchezQ/rusty_results#pydantic

Thanks again!

**Reply by PrettyWood** [reply]
You can create your own json encoder and suggest its usage in the config of pydantic models.
Now if this has to be done purely automatically then it's probably related to https://github.com/samuelcolvin/pydantic/issues/951.
Having custom `__serialize__` and `__deserialize__` (which is already here with `__get_validators__`) methods would be a great thing (as you would implement `Serializer` / `Deserializer` traits with `serde`).
I think the easiest would be to add a check in `pydantic_encoder` and also update the JSON schema in this case. I haven't read the whole issue in details but performances matter so it needs to affect them as less as possible

**Reply by danielSanchezQ** [reply]
We actually had some implementation with the `__get_validators__` that solved our deserialize issues. But not the serialize ones. I also considered providing a custom encoder, but I don't think it is really a proper solution.
It looks like #951 is what we want exactly. But for what I see proposed there it is not internal to pydantic but some tailored solution. So, the target would be for anyone agnostic to pydantic just to use the type within a model and get it properly serialized/deserialized. Should I follow #951 then?

**Reply by danielSanchezQ** [reply]
@PrettyWood I forgot to mention that the type itself is a generic type, and does not inherit from `BaseModel` (hence it is not a model). That may be one source of problems ofc (since it may lack some behavior that `BaseModel` would solve)
